### PR TITLE
Convert to Swift 3

### PR DESCRIPTION
**Pull request by broadwaylamb**
## Pull request checklist
- [x] All tests pass. Demo project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have followed the [coding style](https://github.com/jessesquires/HowToContribute#style-guidelines), and reviewed the [contributing guidelines](https://github.com/jessesquires/HowToContribute).
## What's in this pull request?

This pull request makes the project available for Xcode 8.

I also added testing support for Linux, but didn't configure .travis.yml for it.

@jessesquires please when this is merged can you create a tag so I could use Swift PM to add the project as a dependency?

### CODE REVIEW

File: Tests/LinuxMain.swift [hunk 1]
```diff
@@ -0,0 +1,14 @@
+//
+//  LinuxMain.swift
+//  DefaultStringConvertible
+//
+//  Created by Sergej Jaskiewicz on 13.09.16.
+//  Copyright © 2016 Hexed Bits. All rights reserved.
+//
```
Comment (jessesquires) [inline Tests/LinuxMain.swift:17]: let's update to use project's comment header

File: .travis.yml [hunk 2]
```diff
@@ -11,13 +12,12 @@ env:
...
-   - OSX_SDK=macosx10.11
-   - TVOS_SDK=appletvsimulator9.2
-   - WATCHOS_SDK=watchsimulator2.2
+   - IOS_SDK=iphonesimulator10.0
+   - OSX_SDK=macosx10.12
+   - TVOS_SDK=appletvsimulator10.0
+   - WATCHOS_SDK=watchsimulator3.0
 
    matrix:
-   - DESTINATION="OS=8.1,name=iPhone 4s"      SDK="$IOS_SDK" SCHEME="$IOS_SCHEME" RUN_TESTS="YES" POD_LINT="YES"
```
Comment (jessesquires) [inline .travis.yml:20]: why was this removed? mistake? (we should keep this)

Comment (broadwaylamb) [reply .travis.yml:20]: @jessesquires as far as I know iPhone 4s is not supported anymore.

**Comment by jessesquires** [comment]
Thanks so much @broadwaylamb ! 😍 🎉 Massive help. 

Once travis is green, I'll merge and tag.

> I also added testing support for Linux, but didn't configure .travis.yml for it.

This is awesome, but if we aren't running tests on travis, then we should probably remove it. Otherwise, it's just a maintenance burden.

> when this is merged can you create a tag

Yep! 👍

**Comment by jessesquires** [comment]
@broadwaylamb -- Also, would you like push access to this project? 

I can add you as a contributor. 😄

**Comment by jessesquires** [comment]
Oh, 1 more thing, let's add this to the `CHANGELOG`

```
2.0.0
-----

- Migrated to Swift 3 (#3, @broadwaylamb)

```

File: .travis.yml (hunk 2 above)
Comment (jessesquires) [reply .travis.yml:20]: Hm, should be. It runs iOS 8+. And this lib is iOS 8+

Comment (broadwaylamb) [reply .travis.yml:20]: @jessesquires indeed... I remember Travis didn't support iOS 8 on Xcode 8 a couple of months ago. It seems they have fixed it.

**Comment by broadwaylamb** [comment]
I'd be happy to be a collaborator!

And I'm gonna push those changes tomorrow.

**Comment by jessesquires** [comment]
ping @broadwaylamb 😊

**Comment by broadwaylamb** [comment]
Just pushed some changes. Travis is extremely slow for some reason.

**Comment by broadwaylamb** [comment]
Well, that's weird. When I use `language: swift` in .travis.yml, it builds (kinda) on macOS and does not on Linux. When i use `language: generic` as recommended for building Swift projects on Linux, it builds on Linux, but on macOS I get segfault in Ruby. Have no idea how it all can be connected.

Not to mention build timeout for iOS 8 builds and this

```
xcodebuild: error: Failed to build project DefaultStringConvertible with scheme DefaultStringConvertible-iOS.
    Reason: A build only device cannot be used to run this target.
```

for all the other builds.

And I would like to request some help with CI, because I do not have enough experience for this, obviously.

Yeah, we could remove Linux build at all, but since the project can run on this OS, why not test it on it?

P. S. One last attempt

File: .travis.yml [hunk 3]
```diff
@@ -11,43 +10,74 @@ env:
...
-   - DESTINATION="arch=x86_64" SDK="$OSX_SDK" SCHEME="$OSX_FRAMEWORK_SCHEME" RUN_TESTS="YES" POD_LINT="NO"
-
-   - DESTINATION="OS=9.0,name=Apple TV 1080p" SDK="$TVOS_SDK" SCHEME="$TVOS_SCHEME" RUN_TESTS="YES" POD_LINT="NO"
-
-   - DESTINATION="OS=2.0,name=Apple Watch - 38mm" SDK="$WATCHOS_SDK" SCHEME="$WATCHOS_SCHEME" RUN_TESTS="NO" POD_LINT="NO"
-   - DESTINATION="OS=2.2,name=Apple Watch - 42mm" SDK="$WATCHOS_SDK" SCHEME="$WATCHOS_SCHEME" RUN_TESTS="NO" POD_LINT="NO"
+matrix:
+  include:
+  - language: swift
+    os: osx
```
Comment (jessesquires) [inline .travis.yml:21]: is it possible to just have to `language:` and `os:` once for osx and swift?

**Comment by jessesquires** [comment]
👍 sounds good.

if travis doesn't behave, let's just remove the linux stuff from `.travis` for now and land this 😄

File: .travis.yml (hunk 3 above)
Comment (broadwaylamb) [reply .travis.yml:21]: As far as I know, it is not. Not sure though. However this change helps much.

**Comment by broadwaylamb** [comment]
Well, now it does (last three builds succeeded), so let's keep it ;)

We're only left with this strange xcodebuild error and timeouts on iOS 8.

File: Sources/DefaultStringConvertible.swift [hunk 4]
```diff
@@ -187,28 +187,42 @@ private func generateDeepDescription(any: Any) -> String {
...
         return string + "\r}"
 
-    case .Optional:
+    case .optional:
         return generateDefaultDescription(any)
     }
 }
+
+// Since these methods are not available in Linux
+#if !(os(macOS) || os(iOS) || os(watchOS) || os(tvOS))
```
Comment (jessesquires) [inline Sources/DefaultStringConvertible.swift:217]: can't we say `if os(Linux)` ?

File: Tests/DefaultStringConvertibleTests/DefaultStringConvertibleTests.swift [hunk 5]
```diff
@@ -82,40 +97,49 @@ final class DefaultStringConvertibleTests: XCTestCase {
...
     }
 
     func test_thatClass_providesDeepDescription() {
+
+	// Since in Swift Foundation customMirror for NSDictionary is not yet implemented
+	#if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
         let v = SomeClass(prop1: [1, 2, 3], prop2: [0, "hello world", NSDictionary()])
-        let deepDescription = v.deepDescription
+	#else
+        let v = SomeClass(prop1: [1, 2, 3], prop2: [0, "hello world", [String : String]()])
```
Comment (jessesquires) [inline Tests/DefaultStringConvertibleTests/DefaultStringConvertibleTests.swift:111]: this complexity isn't necessary. let's just have the `let v = ...` in the else block

File: Tests/DefaultStringConvertibleTests/DefaultStringConvertibleTests.swift [hunk 6]
```diff
@@ -198,7 +222,8 @@ class SomeClass: CustomStringConvertible {
     }
 }
 
-
+// Since in Swift Foundation customMirror for NSNumber, NSString and NSArray is not yet implemented
+#if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
```
Comment (jessesquires) [inline Tests/DefaultStringConvertibleTests/DefaultStringConvertibleTests.swift]: let's just remove the use of the `NS*` classes so can we remove this `#if`

File: Tests/DefaultStringConvertibleTests/DefaultStringConvertibleTests.swift [hunk 7]
```diff
@@ -82,40 +97,49 @@ final class DefaultStringConvertibleTests: XCTestCase {
...
+
+	let deepDescription = v.deepDescription
 
         XCTAssertEqual(deepDescription,
                        "<SomeClass> {\r    prop1: [\r        1,\r        2,\r        3\r    ],\r    prop2: "
                         + "[\r        0,\r        \"hello world\",\r        [:]\r    ]\r}")
         print("\n", deepDescription, "\n")
     }
 
+    #if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
```
Comment (jessesquires) [inline Tests/DefaultStringConvertibleTests/DefaultStringConvertibleTests.swift]: with my comment below, we can enable this test on linux, right?

File: Sources/DefaultStringConvertible.swift (hunk 4 above)
Comment (broadwaylamb) [reply Sources/DefaultStringConvertible.swift:217]: We can, but this will decrease the level of generality. Swift sometime will come to platforms other than Linux and macOS, right?

**Comment by jessesquires** [comment]
thanks @broadwaylamb !
- I pushed a couple commits to this PR 
- left some comments about linux

**Comment by broadwaylamb** [comment]
OK, I'll make some changes as soon as I get some free minutes 😊

**Comment by broadwaylamb** [comment]
@jessesquires please take a look at the build. It seems some jobs have failed, but nonetheless are marked as passed. Like [this](https://travis-ci.org/jessesquires/DefaultStringConvertible/jobs/161109373). I'm not sure what causes this.

**Comment by jessesquires** [comment]
Looks good to me. Will go ahead and merge. 👍 💯 

travis is having issues with iOS 8 timing out

**Comment by broadwaylamb** [comment]
@jessesquires yes, but look at [this](https://travis-ci.org/jessesquires/DefaultStringConvertible/jobs/161109373) carefully
